diff --git a/updatesrc/private/updatesrc_update.bzl b/updatesrc/private/updatesrc_update.bzl
index c0cbfc1..a8d24ca 100644
--- a/updatesrc/private/updatesrc_update.bzl
+++ b/updatesrc/private/updatesrc_update.bzl
@@ -49,7 +49,7 @@ if [[ ! -z "${BUILD_WORKSPACE_DIRECTORY}" ]]; then
   cd "${BUILD_WORKSPACE_DIRECTORY}"
 fi
 """ + "\n".join([
-            "cp -f $(readlink \"${{runfiles_dir}}/{out}\") {src}".format(
+            "cp -f $(readlink \"${{runfiles_dir}}/{out}\") {src}\nchmod 0444 {src}".format(
                 src = updsrc.src.short_path,
                 out = updsrc.out.short_path,
             )
